655 Create CollapsiblePanelLayout and Replace Custom Panels with the Shared Component

lizzy-liang-bigstack opened the pull request:
#### What type of PR is this?

- Enhancement

#### What this PR does / why we need it

- ⚠️ This PR is based on #654 

### 1. Create `CollapsiblePanelLayout` [568cedb](https://github.com/bigstack-oss/cube-cos-ui/pull/656/commits/568cedb80d15a74466cb6fff7b940b2ee35a60bc)

``` ts
// Example usage of CosCollapsiblePanelLayout
// The open/close state can be managed internally (uncontrolled)
// or passed down from a parent (controlled).
<CosCollapsiblePanelLayout
  defaultPanelOpen={true} // 👈🏻 uncontrolled example
  // isControlledPanelOpen={isOpen} // 👈🏻 controlled example
  // onControlledPanelOpenChange={setIsOpen} // 👈🏻 controlled example
  rightPanelWidthPercentage={40}
>
  <CosCollapsiblePanelLayout.LeftPanel topic="Add Attributes to Select Events">
    {/* Left panel content */}
    <AttributePanel
      isBuiltIn={isBuiltIn}
      isInitialDataLoading={isInitialDataLoading}
      isMaterialsLoading={isMaterialsLoading}
      isAttributeChanged={isAttributeChanged}
      payload={payload}
      attribute={attribute}
      // ...
    />
  </CosCollapsiblePanelLayout.LeftPanel>

  <CosCollapsiblePanelLayout.RightPanel topic="Attribute Result">
    {/* Right panel content */}
    <AttributeResultPanel
      isMatchingEventsLoading={isMatchingEventsLoading}
      matchingEvents={matchingEvents ?? []}
      // ...
    />
  </CosCollapsiblePanelLayout.RightPanel>
</CosCollapsiblePanelLayout>
```

### 2. Update current UI with the component

- Create / Edit Triggers Page (Select Events) [af1b964](https://github.com/bigstack-oss/cube-cos-ui/pull/656/commits/af1b96496b1ef0fd8089d05396258fb6840f90a0)

[Before]

https://github.com/user-attachments/assets/9aaa2669-5371-4dfb-94c3-71f0f0bdeee4

[After]

https://github.com/user-attachments/assets/2cc10c4e-2ff5-474f-8654-5bbd91491af0

- Node IMPI Page [b388bee](https://github.com/bigstack-oss/cube-cos-ui/pull/656/commits/b388bee83b16f2b8cabfba46d39acc24d0506aa3)


https://github.com/user-attachments/assets/aba8143a-e501-4776-a402-f5ddeac74321


- Storage Integration [e47fbb3](https://github.com/bigstack-oss/cube-cos-ui/pull/656/commits/e47fbb38aecb96e8476b6875227630d0d6f1f2ae)


https://github.com/user-attachments/assets/1b9111da-be9c-4e3e-88a5-a44e90fb15f8


https://github.com/user-attachments/assets/07ad926f-2761-4bac-a66f-ac222b9ebcf1


- Firmware List Pages [962da19](https://github.com/bigstack-oss/cube-cos-ui/pull/656/commits/962da19b9be6de723098dbf8f2bd0075f8f5a651)

https://github.com/user-attachments/assets/9cd5e471-d8cb-4182-95ea-5c8b146087ac

- Fixpack List Page [7a79e04f](https://github.com/bigstack-oss/cube-cos-ui/pull/656/commits/7a79e04f4d3146632c8306a3d42d1e593423c857)

- Health Details Page [44303e2](https://github.com/bigstack-oss/cube-cos-ui/pull/656/commits/44303e206d5ec6b21c15508f2a31efee533ff1e5)

https://github.com/user-attachments/assets/7a3a8878-efca-4f56-8f1b-d78312333490


#### Which issue(s) this PR fixes

Fixes #655 #139

Related to #503 

#### Special notes for your reviewer

> Note

#### Additional documentation

```docs

```

lizzy-liang-bigstack committed:
feat(web-app): create shared collapsiable panels and update current implementation

Signed-off-by: Lizzy Liang <lizzy.liang@bigstack.co>

feat: replace with CosGeneralPanel in events list page

Signed-off-by: Lizzy Liang <lizzy.liang@bigstack.co>

feat: replace custom panel with CosGeneralPanel

Signed-off-by: Lizzy Liang <lizzy.liang@bigstack.co>

feat(web-app): create CollapsiblePanelLayout and related components

Signed-off-by: Lizzy Liang <lizzy.liang@bigstack.co>

feat: update CollapsiblePanels

Signed-off-by: Lizzy Liang <lizzy.liang@bigstack.co>

feat: update Create/Edit Triggers page (Select Events)

Signed-off-by: Lizzy Liang <lizzy.liang@bigstack.co>

feat: update IPMI Control page

Signed-off-by: Lizzy Liang <lizzy.liang@bigstack.co>

feat: update LogConsole

Signed-off-by: Lizzy Liang <lizzy.liang@bigstack.co>

feat: update Create/Edit Storage page

Signed-off-by: Lizzy Liang <lizzy.liang@bigstack.co>

feat: update Firmware List page

Signed-off-by: Lizzy Liang <lizzy.liang@bigstack.co>

feat: update Health Detail page

Signed-off-by: Lizzy Liang <lizzy.liang@bigstack.co>

feat: update Fixpack List page

Signed-off-by: Lizzy Liang <lizzy.liang@bigstack.co>

raven-pan commented on packages/cube-frontend-web-app/src/components/CollapsiblePanelLayout/CollapsiblePanelLayout.tsx:
```diff
@@ -0,0 +1,94 @@
+  })
+
+  // Ensure the panel width percentage stays within 0-100.
+  const clampedWidth = Math.min(100, Math.max(0, rightPanelWidthPercentage))
```
nitpick: we can use the `clamp` function from lodash here.

raven-pan commented on packages/cube-frontend-web-app/src/components/CollapsiblePanelLayout/useCollapsiblePanels.ts:
```diff
@@ -0,0 +1,66 @@
+
+      if (!isControlled) {
+        if (next) {
+          onOpenInternal()
```
We can use the `toggle` function from `useOpenState` to manage this state more easily.

raven-pan commented on packages/cube-frontend-web-app/src/components/CollapsiblePanelLayout/LeftPanel.tsx:
```diff
@@ -0,0 +1,57 @@
+
+  const { toggleOpen } = useContext(CollapsiblePanelsContext)
+
+  const button = isValidElement(customToggleButton) ? (
```
Discussion: Maybe we could use `console.warn` to hint `web-app` developes the `customToggleButton` is not a valid element, and use default button when the `customToggleButton` is `undefined`, What do you think?

lizzy-liang-bigstack replied:
That’s a good idea! 👍 
I’ll go ahead and add this in the update. 
Thanks for the suggestion!

raven-pan commented on packages/cube-frontend-web-app/src/components/CollapsiblePanelLayout/LeftPanel.tsx:
```diff
@@ -0,0 +1,57 @@
+  CosGeneralPanelContentProps,
+  'children' | 'topic' | 'leftSlot' | 'rightSlot'
+> & {
+  customToggleButton?: React.ReactElement<{ onClick?: React.MouseEventHandler }>
```
Do we need to customize the toggleButton?

raven-pan reviewed (commented):
I'm not complete this review yet, but I want to discuss that maybe adding `CollapsiblePanel` to Storybook will make it more maintainable and it will demonstrate to following web-app developers how to use this component.

What do you think?

lizzy-liang-bigstack commented:
Hi @raven-pan, 

Thank you for the feedback! I think it's indeed a great point adding `CollapsiblePanel` to Storybook for better understanding and maintenance. I'll update the PR as soon as it's been done. Thanks again 🙇‍♀️❤️‍🔥

lizzy-liang-bigstack committed:
feat(ui-library): add CollapsiblePanelLayout to Storybook

Signed-off-by: Lizzy Liang <lizzy.liang@bigstack.co>

feat: replace original panel with updated CosCollapsiblePanelLayout

Signed-off-by: Lizzy Liang <lizzy.liang@bigstack.co>

fix: rebase error

Signed-off-by: Lizzy Liang <lizzy.liang@bigstack.co>

fix: comments

Signed-off-by: Lizzy Liang <lizzy.liang@bigstack.co>

raven-pan commented on packages/cube-frontend-ui-library/src/components/CosCollapsiblePanelLayout/useCollapsiblePanels.ts:
```diff
@@ -0,0 +1,50 @@
+
+type UseCollapsiblePanels = {
+  isOpen: boolean
+  toggle: (open: boolean) => void
```
nitpick: We can define the open parameter as `open? boolean`.

raven-pan commented on packages/cube-frontend-ui-library/src/components/CosCollapsiblePanelLayout/LeftPanel.tsx:
```diff
@@ -0,0 +1,68 @@
+    <button
+      type="button"
+      className="inline-flex size-[26px] items-center justify-center rounded-full bg-blue-150"
+      onClick={() => toggle(!isOpen)}
```
Nitpick: We can just call the toggle without any parameter, the toggle function will handle the state change.

raven-pan commented on packages/cube-frontend-ui-library/src/components/CosCollapsiblePanelLayout/LeftPanel.tsx:
```diff
@@ -0,0 +1,68 @@
+
+  let button: React.ReactNode
+
+  if (customToggleButton === undefined) {
```
Just personal style preference, no need to change 🙏

I prefer to extract the render button logic to a function using early return:

```tsx
const renderButton = () => {
  const defaultButton = (
    <button
      type="button"
      className="inline-flex size-[26px] items-center justify-center rounded-full bg-blue-150"
      onClick={() => toggle()}
    >
      <InformationCircle className="icon-lg text-functional-text" />
    </button>
  )

  if (isNil(customToggleButton)) {
    return defaultButton
  }

  if (!isValidElement(customToggleButton)) {
    console.warn(
      'CosCollapsiblePanelLayout.LeftPanel: `customToggleButton` is not a valid React element. Falling back to default button.',
    )
    return defaultButton
  }

  return cloneElement(
    customToggleButton as React.ReactElement<{
      onClick?: React.MouseEventHandler
    }>,
    {
      onClick: (e: React.MouseEvent) => {
        customToggleButton.props.onClick?.(e)
        toggle(!isOpen)
      },
    },
  )
}
```

raven-pan reviewed (approved):
@lizzy-liang-bigstack thanks for your works!

Just some nitpicks and a style issue:

<img width="1390" height="844" alt="Image" src="https://github.com/user-attachments/assets/2a5da174-17ab-4403-945f-1edfe288d8e9" />


Once the style issue is fixed,
feel free to merge, thanks!
